pennylane/data/base/dataset.py

```diff
@@ -0,0 +1,265 @@
+    doc: Optional[str] = None,
+    py_type: Optional[Any] = None,
+    extra: Optional[Dict[str, Any]] = None,
+    default_factory: Optional[Callable[[], T]] = None,
+    **kwargs,
+) -> Any:
+    """Used to define fields on a declarative Dataclass.
+
+    Args:
+        default: See ``Attribute.default``
+        attribute_type: ``AttributeType`` class for this attribute. If not provided,
+            type may be derived from the type annotation on the class.
+        doc: Documentation for the attribute
+        py_type: Type annotation or string describing this object's type. If not
+            provided, the annotation on the class will be used
+        default_factory: See ``Attribute.default_factory``
+    """
+
+    return Attribute(
+        cast(Type[AttributeType[ZarrAny, T, T]], attribute_type),
+        AttributeInfo(doc=doc, py_type=py_type, extra=extra or {}),
+        default=default,
+        default_factory=default_factory,
+    )
+
+
+@dataclass_transform(
+    order_default=False, eq_default=False, field_specifiers=(attribute,), kw_only_default=True
+)
+class _DatasetMeta(type):
+    """This is a metaclass that tells the type system that
+    ``DatasetBase`` behaves like a dataclass. See:
+    https://peps.python.org/pep-0681/
+    """
+
+
+class Dataset(MapperMixin, metaclass=_DatasetMeta):
+    """
+    Base class for Datasets.
+
+    Attributes:
+        fields: A mapping of attribute names to their ``Attribute`` information. Note that
+            this contains attributes declared on the class, not attributes added to
+            an instance. Use ``attrs`` to view all attributes on an instance.
+        bind: The Zarr group that contains this dataset.
+    """
+
+    type_id = "dataset"
+
+    fields: ClassVar[typing.Mapping[str, Attribute]] = MappingProxyType({})
+
+    bind: InitVar[Optional[ZarrGroup]] = attribute(default=None, kw_only=False)  # type: ignore
+
+    def __init__(
+        self,
+        bind: Optional[ZarrGroup] = None,
+        **attrs: Any,
+    ):
+        """
+        Load a dataset from a Zarr Group or initialize a new Dataset.
+
+        Args:
+            bind: The Zarr group, or path to zarr file, that will contain this dataset.
+                If None, the dataset will be stored in memory. Any attributes that
+                already exist in ``bind`` will be loaded into this dataset.
+            **attrs: Attributes to add to this dataset.
+        """
+        if bind is None:
+            self._bind = zarr.group()
+            self._bind.attrs["type_id"] = self.type_id
+        else:
+            self._bind = bind
+
+        self._validate_arguments(attrs)
+        for name, attr in attrs.items():
+            setattr(self, name, attr)
+
+    @property
+    def bind(self) -> ZarrGroup:
+        """The Zarr group that contains this dataset's attributes."""
+        return self._bind
+
+    @property
+    def attrs(self) -> typing.Mapping[str, AttributeType]:
```

Comment:
There's a slight difference between the previous and current output for `attrs`. They don't need to be the same, but the readability of the current version could be improved.
```
>>> dataset = qml.data.Dataset(v1 = 1, v2=2)
>>> dataset.attrs
mappingproxy({'v1': <pennylane.data.attributes.scalar.DatasetScalar object at 0x7f65e485a280>,
              'v2': <pennylane.data.attributes.scalar.DatasetScalar object at 0x7f65e4859ca0>})

>>> dataset = qml.olddata.Dataset(v1 =1, v2=2)
>>> dataset.attrs
{'v1': 1, 'v2': 2}
```

Reply:
Nice catch - I've updated repr to be more informative. It want the type information to be visible, so now it looks like this:
```
>>> Dataset(x=1, y=2).attrs
mappingproxy({'x': DatasetScalar(1), 'y': DatasetScalar(2)})
```